unconstrained_test.go

```diff
@@ -901,130 +1202,117 @@ func TestGradientDescentBacktracking(t *testing.T) {
+	testMinimize(t, tests, &LBFGS{})
 }
 
-func testMinimize(t *testing.T, method Method) {
-	// This should be replaced with a more general testing framework with
-	// a plugable method
-
-	for i, test := range []struct {
-		F Function
-		X []float64
-
-		OptVal float64
-		OptLoc []float64
-
-		Tol      float64
-		Settings *Settings
-	}{
-		{
-			F:      Rosenbrock{},
-			X:      []float64{15, 10},
-			OptVal: 0,
-			OptLoc: []float64{1, 1},
-			Tol:    1e-4,
+func testMinimize(t *testing.T, tests []UnconstrainedTest, method Method) {
+	for _, test := range tests {
+		settings := &Settings{
+			FunctionAbsTol: math.Inf(-1),
+		}
+		if test.gradTol == 0 {
+			settings.GradientAbsTol = 1e-12
+		} else {
+			settings.GradientAbsTol = test.gradTol
+		}
+		if test.tol == 0 {
+			test.tol = 1e-5
+		}
 
-			Settings: DefaultSettings(),
-		},
-		{
-			F:      Rosenbrock{},
-			X:      []float64{-150, 100, 5, -6},
-			OptVal: 0,
-			OptLoc: []float64{1, 1, 1, 1},
-			Tol:    1e-4,
-
-			Settings: &Settings{
-				FunctionAbsTol: math.Inf(-1),
-				GradientAbsTol: 1e-12,
-			},
-		},
-		{
-			F:      Rosenbrock{},
-			X:      []float64{15, 10},
-			OptVal: 0,
-			OptLoc: []float64{1, 1},
-			Tol:    1e-4,
-
-			Settings: &Settings{
-				FunctionAbsTol: math.Inf(-1),
-				GradientAbsTol: 1e-12,
-			},
-		},
-		{
-			F:      Rosenbrock{},
-			X:      []float64{-1.2, 1},
-			OptVal: 0,
-			OptLoc: []float64{1, 1},
-			Tol:    1e-4,
-
-			Settings: &Settings{
-				FunctionAbsTol: math.Inf(-1),
-				GradientAbsTol: 1e-3,
-			},
-		},
-		/*
-			// TODO: Turn this on when we have an adaptive linsearch method.
-			// Gradient descent with backtracking will basically never finish
-			{
-				F:      Linear{8},
-				X:      []float64{9, 8, 7, 6, 5, 4, 3, 2},
-				OptVal: negInf,
-				OptLoc: []float64{negInf, negInf, negInf, negInf, negInf, negInf, negInf, negInf},
-
-				Settings: &Settings{
-					FunctionAbsTol: math.Inf(-1),
-				},
-			},
-		*/
-	} {
-		test.Settings.Recorder = nil
-		result, err := Local(test.F, test.X, test.Settings, method)
+		result, err := Local(test.f, test.x, settings, method)
 		if err != nil {
-			t.Errorf("error finding minimum: %v", err.Error())
+			t.Errorf("error finding minimum (%v) for:\n%v", err, test)
 			continue
 		}
-		// fmt.Println("%#v\n", result) // for debugging
-		// TODO: Better tests
-		if math.Abs(result.F-test.OptVal) > test.Tol {
-			t.Errorf("Case %v: Minimum not found, exited with status: %v. Want: %v, Got: %v", i, result.Status, test.OptVal, result.F)
+
+		if result == nil {
+			t.Errorf("nil result without error for:\n%v", test)
 			continue
 		}
-		if result == nil {
-			t.Errorf("Case %v: nil result without error", i)
+
+		// Check that the optimum function value is as expected.
+		if math.Abs(result.F-test.optVal) > test.tol {
+			t.Errorf("Minimum not found, exited with status: %v. Want: %v, Got: %v for:\n%v",
+				result.Status, test.optVal, result.F, test)
 			continue
 		}
```

Comment:
All things being equal, I'd prefer it stay out, but if it's causing a bunch of trouble I could bend. There are two issues with including it:
1) Location is used for the starting location, and we don't want the user (having to) fill in the information about the gradient. It would be nicer if it was invisible completely, which means making it private
2) If it's private, we need to make sure our optimizers don't actually use it. I like that our optimizers are not privileged -- anyone could re-implement externally the optimizers we have and plug them into Local.

It's not clear to me what the problem is yet and how that's solved by moving the gradient storage location.